### PR TITLE
8331224: ClassCastException in ObjectInputStream hides ClassNotFoundException

### PR DESCRIPTION
**Pull request by RogerRiggs**
The issue reported a ClassCastException "cannot assign instance of java.util.CollSer to field of type java.util.Map"
while deserializing an object referring to an immutable Map that contained a reference to a class that was not available.
Immutable Collections such as Map utilize a serialization proxy in their serialized form.
During deserialization the serialization proxy (a private implementation class) was attempted to be set in a field resulting in the ClassCastException. The ClassCastException and bug hid the ClassCastException that should have been thrown.

When reading record fields or fields of a class, the results of deserialization of individual fields are recorded as dependencies of the object being constructed.
The apparent bug is that the summary of those dependencies is not checked between reading the fields and invoking the constructor to create the record or assigning the fields to an object being constructed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331224](https://bugs.openjdk.org/browse/JDK-8331224): ClassCastException in ObjectInputStream hides ClassNotFoundException (**Bug** - P4)

### Reviewers
 * [Stuart Marks](https://openjdk.org/census#smarks) (@stuart-marks - **Reviewer**) ⚠️ Review applies to [cb9cad62](https://git.openjdk.org/jdk/pull/19043/files/cb9cad6261027b6b83e2f86416c982c6c1780af8)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19043/head:pull/19043` \
`$ git checkout pull/19043`

Update a local copy of the PR: \
`$ git checkout pull/19043` \
`$ git pull https://git.openjdk.org/jdk.git pull/19043/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19043`

View PR using the GUI difftool: \
`$ git pr show -t 19043`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19043.diff">https://git.openjdk.org/jdk/pull/19043.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19043#issuecomment-2088916293)

### CODE REVIEW

**Review by stuart-marks — APPROVED** [review]
OK, I finally went through the changes. First, it's correct to assign the field values only when passHandle is not marked with an exception. It's fairly subtle but if passHandle is marked with an exception, the exception will be thrown by one of the methods farther up the call stack. Second, the test looks pretty comprehensive.

The main difficulty I had is not with the changes here but that the code in this area is handling rather too many cases, including: record/ordinary-class, has vs no special read method, read data into an object vs skipping data, and probably a few other cases. I'm not sure what, if anything, should be done about this. But in any case this should be a subject of a separate conversation.

File: src/java.base/share/classes/java/io/ObjectInputStream.java [hunk 1]
```diff
@@ -2371,6 +2371,9 @@ private Object readRecord(ObjectStreamClass desc) throws IOException {
         }
 
         FieldValues fieldValues = new FieldValues(desc, true);
+        if (handles.lookupException(passHandle) != null) {
+            return null;     // slot marked with exception, don't create record
+        }
```
[the remark this answers is not in the document]
Comment (RogerRiggs) [reply src/java.base/share/classes/java/io/ObjectInputStream.java:2376]: As you discovered, though it returns null at this point, the return from readObject checks for the exception and throws ClassNotFoundException.

**Comment by RogerRiggs** [comment]
Thanks for the review.
I have some ideas to refactor readOrdinaryObject to split out the various subcases and be able to simplify each one.
That's a separate (future) PR.

**Pull request closed without merging**